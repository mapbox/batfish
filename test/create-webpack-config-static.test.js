'use strict';

const path = require('path');
const createWebpackConfigStatic = require('../src/node/create-webpack-config-static');
const createWebpackConfigBase = require('../src/node/create-webpack-config-base');
const validateConfig = require('../src/node/validate-config');
const projectRootSerializer = require('./test-util/project-root-serializer');

jest.mock('../src/node/create-webpack-config-base', () => {
  return jest.fn(() => Promise.resolve({ createdWebpackConfigBase: true }));
});

// Mock mkdirp so validateConfig does not create a temporary directory.
jest.mock('mkdirp', () => ({ sync: jest.fn() }));

expect.addSnapshotSerializer(projectRootSerializer);

function createBatfishConfig(options) {
  return validateConfig(
    Object.assign(
      {
        pagesDirectory: path.join(__dirname, './fixtures/get-pages-data')
      },
      options
    )
  );
}

describe('createWebpackConfigStatic', () => {
  test('creates a base config', () => {
    const batfishConfig = createBatfishConfig();
    return createWebpackConfigStatic(batfishConfig).then((webpackConfig) => {
      expect(createWebpackConfigBase).toHaveBeenCalledTimes(1);
      expect(createWebpackConfigBase).toHaveBeenCalledWith(batfishConfig, {
        target: 'node'
      });
      expect(webpackConfig.createdWebpackConfigBase).toBe(true);
    });
  });

  test('default Webpack config', () => {
    return createWebpackConfigStatic(createBatfishConfig()).then(
      (webpackConfig) => {
        expect(webpackConfig).toMatchSnapshot();
      }
    );
  });

  test('with all Batfish config options that make a difference', () => {
    const batfishConfig = createBatfishConfig({
      webpackStaticIgnore: /stuff\/to\/ignore/,
      webpackStaticStubReactComponent: [
        '/path/to/component/a',
        '/path/to/component/b'
      ],
      webpackConfigStaticTransform: (x) => {
        x.underwentStaticTransform = true;
        return x;
      },
      webpackPlugins: [{ mockPlugin: true }]
    });
    return createWebpackConfigStatic(batfishConfig).then((webpackConfig) => {
      expect(webpackConfig).toMatchSnapshot();
    });
  });
});
