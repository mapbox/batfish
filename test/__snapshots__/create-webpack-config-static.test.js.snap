// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createWebpackConfigStatic default Webpack config 1`] = `
Object {
  "createdWebpackConfigBase": true,
  "devtool": "source-map",
  "entry": Object {
    "static": "<PROJECT_ROOT>/src/webpack/static-render-pages.js",
  },
  "externals": Object {
    "@mapbox/link-hijacker": "<PROJECT_ROOT>/node_modules/@mapbox/link-hijacker/index.js",
    "@mapbox/link-to-location": "<PROJECT_ROOT>/node_modules/@mapbox/link-to-location/index.js",
    "@mapbox/scroll-restorer": "<PROJECT_ROOT>/node_modules/@mapbox/scroll-restorer/index.js",
    "electron": "electron",
    "mkdirp": "<PROJECT_ROOT>/node_modules/mkdirp/index.js",
    "pify": "<PROJECT_ROOT>/node_modules/pify/index.js",
    "react": "react",
    "react-dom": "react-dom",
    "react-dom/server": "react-dom/server",
    "react-helmet": "react-helmet",
    "source-map-support/register": "<PROJECT_ROOT>/node_modules/source-map-support/register.js",
    "uglify-js": "<PROJECT_ROOT>/node_modules/uglify-js/tools/node.js",
  },
  "node": Object {
    "Buffer": false,
    "__dirname": false,
    "__filename": false,
    "console": false,
    "global": false,
    "process": false,
    "setImmediate": false,
  },
  "output": Object {
    "filename": "./static-render-pages.js",
    "libraryTarget": "commonjs2",
  },
  "plugins": Array [
    LimitChunkCountPlugin {
      "options": Object {
        "maxChunks": 1,
      },
    },
  ],
  "target": "node",
}
`;

exports[`createWebpackConfigStatic with all Batfish config options that make a difference 1`] = `
Object {
  "createdWebpackConfigBase": true,
  "devtool": "source-map",
  "entry": Object {
    "static": "<PROJECT_ROOT>/src/webpack/static-render-pages.js",
  },
  "externals": Object {
    "@mapbox/link-hijacker": "<PROJECT_ROOT>/node_modules/@mapbox/link-hijacker/index.js",
    "@mapbox/link-to-location": "<PROJECT_ROOT>/node_modules/@mapbox/link-to-location/index.js",
    "@mapbox/scroll-restorer": "<PROJECT_ROOT>/node_modules/@mapbox/scroll-restorer/index.js",
    "electron": "electron",
    "mkdirp": "<PROJECT_ROOT>/node_modules/mkdirp/index.js",
    "pify": "<PROJECT_ROOT>/node_modules/pify/index.js",
    "react": "react",
    "react-dom": "react-dom",
    "react-dom/server": "react-dom/server",
    "react-helmet": "react-helmet",
    "source-map-support/register": "<PROJECT_ROOT>/node_modules/source-map-support/register.js",
    "uglify-js": "<PROJECT_ROOT>/node_modules/uglify-js/tools/node.js",
  },
  "module": Object {
    "rules": Array [
      Object {
        "loader": "ignore-loader",
        "test": /stuff\\\\/to\\\\/ignore/,
      },
    ],
  },
  "node": Object {
    "Buffer": false,
    "__dirname": false,
    "__filename": false,
    "console": false,
    "global": false,
    "process": false,
    "setImmediate": false,
  },
  "output": Object {
    "filename": "./static-render-pages.js",
    "libraryTarget": "commonjs2",
  },
  "plugins": Array [
    LimitChunkCountPlugin {
      "options": Object {
        "maxChunks": 1,
      },
    },
    NormalModuleReplacementPlugin {
      "newResource": "<PROJECT_ROOT>/src/node/static-stubs/react-component.js",
      "resourceRegExp": /\\\\/path\\\\/to\\\\/component\\\\/a/,
    },
    NormalModuleReplacementPlugin {
      "newResource": "<PROJECT_ROOT>/src/node/static-stubs/react-component.js",
      "resourceRegExp": /\\\\/path\\\\/to\\\\/component\\\\/b/,
    },
  ],
  "target": "node",
  "underwentStaticTransform": true,
}
`;
